### DISCUSSION TITLE
static routes need their root directory on startup

### DISCUSSION THREAD

**Discussion by Pike** [aio-libs/aiohttp#5379] [Q&A]
Hi, I'm having a little pet project where I want to serve data from multiple discs, including USB drives.

https://github.com/aio-libs/aiohttp/blob/3edc43c1bb718b01a1fbd67b01937cff9058e437/aiohttp/web_urldispatcher.py#L537-L538 errors when the drives aren't mounted. And does so in a way that's non-trivial to work around.

I've tried to track down if that's just convenience or for security-related functionality, but that code changed a number of times, and I've lost track.

I wonder if it'd be OK to just drop that check?

**Comment by Raphasha27** [comment]
Static routes in aiohttp serve files from a directory, and that directory must exist at the time the route is registered, not when a request comes in:

```python
import os
from aiohttp import web

app = web.Application()

# This fails if /var/data/static doesn't exist at startup
app.router.add_static("/static", "/var/data/static")
```

**Solutions:**

1. **Ensure the directory exists before starting the app:**
```python
os.makedirs("/var/data/static", exist_ok=True)
app.router.add_static("/static", "/var/data/static")
```

2. **Serve files dynamically with a custom handler** (if the directory is created later):
```python
async def serve_static(request):
    path = os.path.join("/var/data/static", request.match_info["filename"])
    if not os.path.exists(path):
        raise web.HTTPNotFound()
    return web.FileResponse(path)

app.router.add_get("/static/{filename:.*}", serve_static)
```

3. **Use `aiohttp_static` for deferred resolution**:
```python
# pip install aiohttp-static
from aiohttp_static import static_resource

app.router.register_resource(static_resource("/static", "/var/data/static"))
```

The create-server script you're using should ensure the directory exists before calling `web.run_app()`, or use solution #2 for directories that get created at runtime.
